include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h

```diff
@@ -136,47 +236,253 @@ struct __parallel_merge_submitter<_IdType, __internal::__optional_kernel_name<_N
-                __serial_merge(__rng1, __rng2, __rng3, __start.first, __start.second, __i_elem, __chunk, __n1, __n2,
-                               __comp);
+        auto __event = __exec.queue().submit(
+            [&__rng1, &__rng2, &__rng3, __comp, __chunk, __steps, __n1, __n2](sycl::handler& __cgh) {
+                oneapi::dpl::__ranges::__require_access(__cgh, __rng1, __rng2, __rng3);
+                __cgh.parallel_for<_MergeKernelName...>(
+                    sycl::range</*dim=*/1>(__steps), [=](sycl::item</*dim=*/1> __item_id) {
+                        const _IdType __i_elem = __item_id.get_linear_id() * __chunk;
+                        const auto __start = __find_start_point(__rng1, __rng2, __i_elem, __n1, __n2, __comp);
+                        __serial_merge(__rng1, __rng2, __rng3, __start.first, __start.second, __i_elem, __chunk, __n1,
+                                       __n2, __comp);
+                    });
             });
+        // We should return the same thing in the second param of __future for compatibility
+        // with the returning value in __parallel_merge_submitter_large::operator()
+        return __future(__event, std::shared_ptr<__result_and_scratch_storage_base>{});
+    }
+};
+
+template <typename _IdType, typename _CustomName, typename _DiagonalsKernelName, typename _MergeKernelName>
+struct __parallel_merge_submitter_large;
+
+template <typename _IdType, typename _CustomName, typename... _DiagonalsKernelName, typename... _MergeKernelName>
+struct __parallel_merge_submitter_large<_IdType, _CustomName,
+                                        __internal::__optional_kernel_name<_DiagonalsKernelName...>,
+                                        __internal::__optional_kernel_name<_MergeKernelName...>>
+{
+  private:
+    struct nd_range_params
+    {
+        std::size_t base_diag_count = 0;
+        std::size_t steps_between_two_base_diags = 0;
+        _IdType chunk = 0;
+        _IdType steps = 0;
+    };
+
+    // Calculate nd-range parameters
+    template <typename _ExecutionPolicy, typename _Range1, typename _Range2>
+    nd_range_params
+    eval_nd_range_params(_ExecutionPolicy&& __exec, const _Range1& __rng1, const _Range2& __rng2) const
+    {
+        using _Range1ValueType = oneapi::dpl::__internal::__value_t<_Range1>;
+        using _Range2ValueType = oneapi::dpl::__internal::__value_t<_Range2>;
+        using _RangeValueType = std::conditional_t<(sizeof(_Range1ValueType) > sizeof(_Range2ValueType)),
+                                                   _Range1ValueType, _Range2ValueType>;
+
+        const std::size_t __n = __rng1.size() + __rng2.size();
+
+        // Empirical number of values to process per work-item
+        const std::uint8_t __chunk = __exec.queue().get_device().is_cpu() ? 128 : 4;
+
+        const _IdType __steps = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __chunk);
+        const _IdType __base_diag_count = 32 * 1'024;
```

Comment:
Similar to our previous discussion with SLM loading between base diagonals, it seems to me that the ideal distance between base diagonals probably has some minimum to provide enough work for the partitioning kernel, but also probably should be tied to the value of `__n`, to optimize the second kernel.  I would expect that there is some explicit fixed distance between base diagonals which is ideal (for a given hardware).  Right now, the distance between base diagonals grows with `__n` since we have a fixed number.    

I guess as written, this could be `constexpr`, and if it is important for optimization reasons for it to be a known quantity, we could perhaps adjust it at compile time based upon `_IdType`? 

At least we have some ballpark understanding of size based upon the type we are using for ids.

This isn't required, but if it makes sense, lets list this as a issue / TODO for the future.

Reply:
TODO applied for now. Probably will take a look again after merge into main branch.